### ISSUE TITLE
Image Loading Pipeline Enhancements

### ISSUE THREAD

**Issue by slembcke** [cocos2d/cocos2d-objc#480]
Currently the "normal" path for loading a texture from an image is to call [CCTextureCache addImage:] with a filename. That filename is resolved to a device specific filename using [CCFileUtils fullPathFromRelativePath:contentScale:]. [CCTextureCache addImage:] then uses that path to load a CGImage from a UIImage and passes that on to the CCTexture initializer. CCTexture turns the CGImage into a bitmap and creates a GL texture from it. This is a little messy, and there are a few issues.
- The texture cache is responsible for loading image files, which is an odd assignment of responsibility.
- Images are loaded using UIKit which is problematic since it performs it's own suffix substitution outside of Cocos' control. Loading directly to a CGImage makes more sense.
- Images are loaded directly using a file path instead of a stream object. It's not possible to support compressed or encrypted streams.
- There isn't really a discrete step where an image is loaded into a bitmap. This makes it hard to add additional image loading routines.
- CCTextures are created directly from bitmap data or an image object. We need to insert a "CCTextureInfo" object in there somewhere in order clean up the way that caching works.

Here is my improved idea sketched out in code. I apologize for the length of the post... I tried to explain it in a paragraph but that was really hard to follow. The basic ideas are to split up textures in a way to use them with @Birkemose's cache class, and split up file handling (using streams) from bitmap loading and texture uploading.

```
@implementation CCTexture

+(instancetype)textureWithImageNamed:(NSString *)imageName
{
    CCTexture *cached = [TEXTURE_CACHE objectForKey:imageName];
    if(cached){
        return cached;
    } else {
        // A CCAsset would contain wrap info like the resolved and original paths (or URLs) and contentScale.
        CCAssetInfo *asset = [[CCFileUtils sharedFileUtils] assetForFilename:imageName];

        CCTexture *texture = nil;
        if(asset.isPVR){
            // PVR files don't necessarily load a bitmap, need to handle them specially still.
            texture = [CCTextureInfo pvrTextureInfoForAsset:asset];
        } else {
            // Load the bitmap and the texture.
            CCBitmapInfo *bitmap = [CCBitmapInfo bitmapInfoFromAsset:asset];
            texture = [CCTextureInfo textureInfoFromBitmap:bitmap];
        }

        // Cache the texture data using Birkmose's data/object cache class.
        // This will init a CCTexture from the CCTextureInfo when requested from the cache.
        [TEXTURE_CACHE addData:texture forKey:imageName];

        // Now we can just retrive it from the cache again and it will create the CCTexture object.
        return [TEXTURE_CACHE objectForKey:imageName];
    }
}

@end
```

The CCAssetInfo objects would encapsulate the CCFileUtils information and hold onto more of it. They would also transparently handle compressed or encrypted assets using streams:

```
@interface CCAssetInfo

// Original unresolved path.
@property(nonatomic, readonly) NSString *path;

// Resolved (device specific) path.
@property(nonatomic, readonly) NSString *resolvedPath;

// Extension excluding .gz or .ccp (my proposed encrypted file extension)
@property(nonatomic, readonly) NSString *extension;

// Get a stream to read the file. Automatically handle compressed or encrypted files.
// It would be very easy to make subclasses of NSInputStream to support this.
-(NSInputStream *)getStream;

@end
```

Next up CCBitmapObjects would encapsulate the process of loading bitmaps from images of various sorts from streams. It would also handle downscaling, pixel format conversion, and image padding for POT textures. We could also split metadata loading (like size) from the bitmap loading. This is useful because spriteframes lazily load their associated textures, but do need texture size information to set up the texture rects correctly. In the future this would also be useful for loading spritesheets at runtime with efficient memory.

```
@interface CCBitmapInfo : NSObject

// Asset the bitmap is loaded from.
@property(nonatomic, readonly) CCAssetInfo *asset;

// Original pixel size of the image.
@property(nonatomic, readonly) CGSize pixelSize;

// Size of the image after loading. (downscaling may apply)
@property(nonatomic, readonly) CGSize loadedSize;

@property(nonatomic, readonly) GLenum pixelFormat;

-(NSData *)loadBitmap;

@end

@implementation CCBitmapInfo

-(NSData *)loadBitmap
{
    NSString *ext = _asset.extension;
    NSInputStream *stream = [_asset getStream];
    NSUInteger downscaleFactor = DecideDownscalingRulesHere(asset);


    if([ext isEqualToString:@"png"]){
        return [CCBitmapInfo bitmapForPNGStream:stream downscaleFactor:downscaleFactor];
    } else if(...){
        // Handle JPEG, PDF, ImageIO, WebP, etc.
        ...
    }
}

@end
```

Lastly, the CCTextureInfo object would be a rather boring class like GLKTextureInfo that would just wrap up the bitmap data and GL texture name. This would be the "data" object that @Birkemose's cache class would use to create a CCTexture object. Texture objects would become fairly boring wrappers of CCTextureInfo objects.

```
@interface CCTextureInfo

// access to the bitmap info.
@property(nonatomic, readonly) CCBitmapInfo *bitmap;

// GL texture name.
@property(nonatomic, readonly) GLuint name;

@end
```

**Comment by Birkemose** [comment]
I also think this issue has a lot to do with cleaning up CCFileUtils. In general, a lot of people have a hard time understanding how all this works.
I would like to propose a “One Size Fits All”, texture scheme, where a user never will have to create more than one set of images. He can then tell Cocos2D that this set of images is -ipad, and Cocos2D will automatically scale for any other device. This shall of course be backwards compatible, and if an -ipadhd image is found, this will be used in stead. I think @scott already implemented the core in this, we actually only have to add some wrapping.

When loading files, single files, folders and sprite sheets should be 100% interchangeable. The suffix should be 100% flexible, without having to change any setup. I have already provided the base code for this, and it works excellent.

Caching objects should still be possible. The cache class I wrote works great for this. I also have a pool brother class, which based on the same idea as the cache class, can pool generic objects.
Besides the cache scheme, I would like to implement a scratch-pad texture cache. See #525
